### DISCUSSION TITLE
Content Preview

### DISCUSSION THREAD

**Discussion by bluebill1049** [facebook/lexical#1825] [Q&A] [answered]
Thanks for the awesome work.

I was wondering what's the best way to generate a preview? 

```diff
<RichTextPlugin
  placeholder={null}
-  contentEditable={<ContentEditable className={styles.editorInput} />}
+  contentEditable={null} // so it will just render the content
/>
```

is it possible for `RichTextPlugin` to render the user input only?

**Comment by trueadm** [comment, marked as answer]
You can `readOnly` on the editor itself to be `true`. This can be done on `initialConfig` passed to `LexicalComposer` too, or by doing `editor.setReadOnly(true)`.

**Reply by bluebill1049** [reply]
thanks @trueadm it works with the preview, however, for some reason, each keystroke lost its focus. Do i miss use the API below: 

here is a codesandbox: https://codesandbox.io/s/lexical-plain-text-example-forked-r71kh4?file=/src/App.js

```tsx
// rich text
const onChange = (editorState) => {
  if (props.onChange) {
    editorState.read(() => {
      props.onChange(JSON.stringify(editorState))
    })
  }
}

<RichTextPlugin
  contentEditable={<ContentEditable className={styles.editorInput} />}
  placeholder={<Placeholder />}
/>
```

```tsx
// rich text with preview
function PreviewPlugin(props) {
  const [editor] = useLexicalComposerContext()

  useEffect(() => {
    editor.update(() => {
      if (props.value) {
        editor.setEditorState(editor.parseEditorState(props.value))
      }
    })
  }, [editor, props.value])

  return null
}

const editorConfig = {
  theme,
  readOnly: true,
  nodes: [...],
}

export function RichTextPreview(props: Props) {
  return (
    <LexicalComposer initialConfig={editorConfig}>
      <PreviewPlugin value={props.value} />
      <RichTextPlugin
        placeholder={null}
        contentEditable={<ContentEditable className={styles.editorInput} />}
      />
    </LexicalComposer>
  )
}
```

**Reply by trueadm** [reply]
You want to avoid that pattern where possible. That's how DraftJS worked, but it is actually really inefficient.

Instead, you want to serialize your state to JSON, only when you need to be extracted out. You definitely don't want to be forcing the editor state to be a new version on each update, as you'll be killing performance. You can think of `editor.setEditorState` like a load game sort of case – you don't want to be loading game each time you move your character. Lexical's EditorState wasn't designed to be controlled by React, it was designed to be controlled by Lexical. Trust Lexical to handle it properly.

In the case where you only want to show the editor as read only, you can just do what you're already doing and setting `readOnly` to `true`. I believe @zurfyx has a good example of this he sent to someone else asking a similar question recently.

**Reply by bluebill1049** [reply]
thanks again @trueadm that make sense, would be good to see an example where user can see the live updated content. I have looked through the discussion questions, couldn't find a relevant example? if you don't mind @zurfyx would you mind sharing it again? thanks.

**Reply by trueadm** [reply]
You can do something like `editor.setEditorState(editor.parseEditorState(props.value))` to update Lexical, but you should only need to call this when the content has changed from some backend store or something? Can you explain maybe what you're trying to build?

If what you want is for Lexical to switch between being editable and non-editable, you can use `editor.setReadOnly(true/false)`

**Reply by bluebill1049** [reply]
![Screen Shot 2022-04-18 at 7 57 58 pm](https://user-images.githubusercontent.com/10513364/163792144-94d1d856-1f40-4357-906f-c51a9179e935.png)

I would like users to visualize what they have entered on a device (mobile-first) view. I have pretty much got Lex working, thanks for the awesome API design and plugin system. This is the last piece I am seeking some help with before I can deprecate the existing rich text editor.

> You can do something like editor.setEditorState(editor.parseEditorState(props.value)) to update Lexical,

```tsx
useEffect(() => {
  editor.setEditorState(editor.parseEditorState(props.value))
}, [editor, props.value])
```

I am already doing the above, however, I lost the focus on each keystroke.

**Reply by zurfyx** [reply]
You're almost there! So basically you're trying to replicate the content. You can do this with Lexical but there's no built-in way to do this in such efficient manner (efficient diffing) like a single editor.

```
editor.addUpdateListener(({editorState, dirtyLeaves, dirtyElements}) => {
  // Discard selection changes, they're useless for your rendering purposes. Selection = caret
  if (dirtyLeaves.size === 0 && dirtyElements.size === 0) {
   return;
  }
  editor2.setEditorState(editorState);
});
```

The above will work but as @trueadm mentioned you have to be careful with performance. You probably don't want/need to trigger this immediately after every key press. Use debounce or throttle -> Google `web debounce` if you've never seen this technique before 

Lastly, do you need a render view on a WYSIWYG editor? UX wise a render view works wonders when you're typing MarkDown since you want to double check you've typed the correct MarkDown syntax but a WYSIWYG editor should already display the final result.

**Reply by bluebill1049** [reply]
> Lastly, do you need a render view on a WYSIWYG editor? UX wise a render view works wonders when you're typing MarkDown since you want to double check you've typed the correct MarkDown syntax but a WYSIWYG editor should already display the final result.

That's a valid point, perhaps what I should do is once the user commits the change and update the view layer and stop pursuing the real-time preview. Thanks, @trueadm and @zurfyx's help 🙏  have a good Easter break.

**Reply by trueadm** [reply]
Thanks for the explanation as to what you're doing, that really helped. I think you're fine keeping it real-time. You also don't need to stringify the EditorState if you're just passing it to another editor. You can just do

```
editor.addUpdateListener(({editorState, dirtyLeaves, dirtyElements}) => {
  // Discard selection changes, they're useless for your rendering purposes. Selection = caret
  if (dirtyLeaves.size === 0 && dirtyElements.size === 0) {
   return;
  }
  debounce(() => {
    editor2.setEditorState(editorState);
  })
});
```

This should be fine for performance. If you're noticing any strange issues, it might be because the read only editor is trying to update the DOM selection when it reconciles. Thinking about this, it should probably skip all that work if in read only mode as it doesn't do that right now! That might help your issue too?

**Reply by bluebill1049** [reply]
> it should probably skip all that work if in read only mode as it doesn't do that right now!

yea, that makes sense. 

Thanks @trueadm I will give it a try tmr. Got me excited to try it out again. 💪🏻

**Reply by bluebill1049** [reply]
Instead of using `editor.addEventlistner`, could I do the following: 

```tsx
// value will be coming from the main editor
const Update = ({ value }) => {
  const [editor2] = useLexicalComposerContext();

  useEffect(() => {
    editor2.setEditorState(value);
  }, [editor, value]);

  return null;
};
```

I am still experiencing lost focus on keystrokes.

https://codesandbox.io/s/lexical-plain-text-example-forked-r71kh4?file=/src/App.js:383-559

**Reply by zurfyx** [reply]
Ah, that's because selection needs to be set to null, otherwise it tries to replicate it to the other (it's part of EditorState).

You can do `value._selection = null` for now (bad because it should be treated as immutable but will work for now), we should probably introduce a flag to drop selection from the EditorState.

Additionally, you don't want to use `useState` to pass the new EditorState around, it causes a re-render. Instead, you can store a function on a `useCallback` or similar than does the `updateEditor2State` for you directly

**Reply by zurfyx** [reply]
Published a proposal that should work for your use case - https://github.com/facebook/lexical/pull/1848

**Reply by trueadm** [reply]
Ah, you can apply an editorState without selection by cloning it. The clone method lets you specific a new selection for the cases where you want to change, so in this case you can have `null`. Does this work?

`const editorStateWithoutSelection = editorState.clone(null)`

**Reply by bluebill1049** [reply]
> const editorStateWithoutSelection = editorState.clone(null)

I will give that a try 👍 thanks, everyone. The new flag proposal looks awesome as well.

**Reply by bluebill1049** [reply]
@trueadm I can confirm it's working as expected. 👯 I am unblocked.

**Reply by zurfyx** [reply]
Ah yes, let's stick to `clone` then! Glad to hear you got it working!

**Reply by mrkchang** [reply]
@bluebill1049 hey thanks for creating this thread. would you mind sharing your final example? I'm confused on your example using states while the developers here seem to recommend against it due to re-rendering...

Also did you end up registering a listener?

**Reply by Raff-project10x** [reply]
Does this solution work when using RichTextPlugin and edtiable set to false? I have a similar need as preview but to display instead. For example, admin writes blog in editable rich text editor and saves json to db. User sees only the rendered editor html without being able to edit. I am able to load the editor state just fine but nothing is displayed when editable is false.
